packages/flutter_tools/lib/src/device.dart

```diff
@@ -210,103 +275,133 @@ abstract class DeviceManager {
-
-        // Ephemeral is nullable for device types where this is not well
-        // defined.
-        final List<Device> ephemeralDevices = <Device>[
-          for (final Device device in devices)
-            if (device.ephemeral == true)
-              device,
-        ];
-
-        if (ephemeralDevices.length == 1) {
-          devices = ephemeralDevices;
-        }
-      }
+  final FlutterProject? flutterProject;
+  final bool mustBeSupportedByFlutter;
+  final bool mustBeSupportedForProject;
+  final bool mustBeSupportedForAll;
+
+  Future<bool> matchesRequirements(Device device) async {
+    if ((mustBeSupportedByFlutter && !device.isSupported()) ||
+        (mustBeSupportedForProject && !isDeviceSupportedForProject(device)) ||
+        (mustBeSupportedForAll && !await isDeviceSupportedForAll(device))) {
+      return false;
     }
+    return true;
+  }
 
-    return devices;
+  Future<bool> isDeviceSupportedForAll(Device device) async {
+    // User has specified `--device all`.
+    //
+    // Always remove web and fuchsia devices from `--all`. This setting
+    // currently requires devices to share a frontend_server and resident
+    // runner instance. Both web and fuchsia require differently configured
+    // compilers, and web requires an entirely different resident runner.
+    final TargetPlatform devicePlatform = await device.targetPlatform;
+    if (device.isSupported() &&
+        devicePlatform != TargetPlatform.fuchsia_arm64 &&
+        devicePlatform != TargetPlatform.fuchsia_x64 &&
+        devicePlatform != TargetPlatform.web_javascript &&
+        isDeviceSupportedForProject(device)) {
+      return true;
+    }
+    return false;
   }
 
   /// Returns whether the device is supported for the project.
   ///
-  /// This exists to allow the check to be overridden for google3 clients. If
+  /// A device can be supported by Flutter but not supported for the project
+  /// (e.g. when the user has removed the iOS directory from their project).
+  ///
+  /// This also exists to allow the check to be overridden for google3 clients. If
   /// [flutterProject] is null then return true.
-  bool isDeviceSupportedForProject(Device device, FlutterProject? flutterProject) {
+  bool isDeviceSupportedForProject(Device device) {
+    if (!device.isSupported()) {
+      return false;
+    }
     if (flutterProject == null) {
       return true;
     }
-    return device.isSupportedForProject(flutterProject);
+    return device.isSupportedForProject(flutterProject!);
+  }
+}
+
+/// A class for filtering devices.
+///
+/// If [mustBeConnected] is true, only devices detected as connected will be included.
+///
+/// If [supportFilter] is provided, only devices matching the requirements will be included.
+///
+/// If [deviceConnectionFilter] is provided, only devices matching the DeviceConnectionInterface will be included.
+class DeviceDiscoveryFilter {
+  DeviceDiscoveryFilter({
```

Comment:
Nvm, I thought we could make this const, but `DeviceDiscoverySupportFilter` has a `FlutterProject`, which can't be made const.